packages/connection/src/class-error-handler.php

```diff
@@ -0,0 +1,568 @@
+			$stored_error = $this->store_error( $error );
+			if ( $stored_error ) {
+				$this->send_error_to_wpcom( $stored_error );
+			}
+		}
+	}
+
+	/**
+	 * Checks the status of the gate
+	 *
+	 * This protects the site (and WPCOM) against over loads.
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param \WP_Error $error the error object.
+	 * @return boolean $should_report True if gate is open and the error should be reported.
+	 */
+	public function should_report_error( \WP_Error $error ) {
+
+		if ( defined( 'JETPACK_DEV_DEBUG' ) && JETPACK_DEV_DEBUG ) {
+			return true;
+		}
+
+		$bypass_gate = apply_filters( 'jetpack_connection_bypass_error_reporting_gate', false );
+		if ( true === $bypass_gate ) {
+			return true;
+		}
+
+		$transient = self::ERROR_REPORTING_GATE . $error->get_error_code();
+
+		if ( get_transient( $transient ) ) {
+			return false;
+		}
+
+		set_transient( $transient, true, HOUR_IN_SECONDS );
+		return true;
+	}
+
+	/**
+	 * Stores the error in the database so we know there is an issue and can inform the user
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param \WP_Error $error the error object.
+	 * @return boolean|array False if stored errors were not updated and the error array if it was successfully stored.
+	 */
+	public function store_error( \WP_Error $error ) {
+
+		$stored_errors = $this->get_stored_errors();
+		$error_array   = $this->wp_error_to_array( $error );
+		$error_code    = $error->get_error_code();
+		$user_id       = $error_array['user_id'];
+
+		if ( ! isset( $stored_errors[ $error_code ] ) || ! is_array( $stored_errors[ $error_code ] ) ) {
+			$stored_errors[ $error_code ] = array();
+		}
+
+		$stored_errors[ $error_code ][ $user_id ] = $error_array;
+
+		// Let's store a maximum of 5 different user ids for each error code.
+		if ( count( $stored_errors[ $error_code ] ) > 5 ) {
+			// array_shift will destroy keys here because they are numeric, so manually remove first item.
+			$keys = array_keys( $stored_errors[ $error_code ] );
+			unset( $stored_errors[ $error_code ][ $keys[0] ] );
+		}
+
+		if ( update_option( self::STORED_ERRORS_OPTION, $stored_errors ) ) {
+			return $error_array;
+		}
+
+		return false;
+
+	}
+
+	/**
+	 * Converts a WP_Error object in the array representation we store in the database
+	 *
+	 * @since 8.7.0
+	 *
+	 * @param \WP_Error $error the error object.
+	 * @return boolean|array False if error is invalid or the error array
+	 */
+	public function wp_error_to_array( \WP_Error $error ) {
+
+		$data       = $error->get_error_data();
+		$error_code = $error->get_error_code();
+
+		if ( ! isset( $data['signature_details'] ) || ! is_array( $data['signature_details'] ) ) {
+			return false;
+		}
+
+		$data = $data['signature_details'];
+
+		if ( ! isset( $data['token'] ) || empty( $data['token'] ) ) {
+			return false;
+		}
+
+		$user_id = $this->get_user_id_from_token( $data['token'] );
+
+		$error_array = array(
+			'error_code'    => $error_code,
```

Comment:
It looks like this is the only use of the `$error_code` variable in this method. Could we just call `$error->get_error_code()` here?

Reply:
yep, https://github.com/Automattic/jetpack/pull/16194/commits/33a6661d0a83b31cabae0ca6abf3e8d004d9fb55